Trad carol (#13)
appliedepi/epirhandbook_pt: new_pages/basics.Rmd

#### RTools {.unnumbered}

RTools é uma coleção de softwares para construção de pacotes para R

Instale desse website: <https://cran.r-project.org/bin/windows/Rtools/>

#### phantomjs {.unnumbered}

Esta ferramenta é freqüentemente usada para tirar "fotografias da tela" de páginas da web. Por exemplo, quando você cria uma cadeia de transmissão com o pacote **epicontacts**,  um arquivo HTML é produzido e ele é interativo e dinâmico. Se você quiser uma imagem estática, pode ser útil usar o pacote [**webshot**](https://wch.github.io/webshot/articles/intro.html) para automatizar este processo. Isto exigirá o programa externo "phantomjs". Você pode instalar o phantomjs através do pacote **webshot** com o comando `webshot::install_phantomjs()`.

<!-- ======================================================= -->

## RStudio {#rstudio}

### Se orientando no RStudio  {.unnumbered}

> **Primeiramente, abra o RStudio.** Como os ícones deles podem ser muito parecidos, certifique-se de que você está abrindo *RStudio* e não R.

Author: ```suggestion
**Primeiramente, abra o RStudio.** Como os ícones deles podem ser muito parecidos, certifique-se de que você está abrindo mesmo  o *RStudio* e não R.
```

> **Primeiramente, abra o RStudio.** Como os ícones deles podem ser muito parecidos, certifique-se de que você está abrindo mesmo  o *RStudio* e não R.

Author: ```suggestion
**Primeiramente, abra o RStudio.** Como os ícones deles podem ser muito parecidos, certifique-se de que você está abrindo mesmo o *RStudio* e não R.
```

**RStudio** é uma interface (GUI) para facilitar o uso de **R**. Você pode pensar em R como sendo o motor de um veículo, fazendo o trabalho crucial, e RStudio como a carroceria do veículo (com assentos, acessórios, etc.) que ajuda você realmente a usar o motor para seguir em frente! Você pode ver a ficha completa da interface de usuário do RStudio (PDF) [aqui](https://www.rstudio.com/wp-content/uploads/2016/01/rstudio-IDE-cheatsheet.pdf)

Por padrão, o RStudio exibe quatro painéis retangulares.

```{r out.width = "100%", fig.align = "center", echo=F}
knitr::include_graphics(here::here("images", "RStudio_overview.png"))
```

[***DICA:*** Se o seu RStudio mostrar apenas um painel esquerdo, é porque não há nenhum script aberto ainda]{style="color: black;"}

**O Painel Fonte**
Este painel, por padrão localizado na parte superior esquerda, é um espaço onde é possível editar, executar e salvar seus [scripts](#scripts). Os scripts contêm os comandos que você deseja executar. Este painel também pode exibir conjuntos de dados (quadros de dados) para visualização.

Para usuários da Stata, este painel é similar às janelas de seu Do-file e Data Editor.

**O Painel de Console R**

> O Console R, ou simplesmente Console, é por padrão o painel esquerdo ou inferior esquerdo do RStudio. Ele é a casa do "motor" R. É aqui que os comandos são realmente executados e as saídas não gráficas e as mensagens de erro/aviso aparecem. Você pode digitar e executar comandos diretamente no Console, mas perceba que estes comandos não são salvos como são ao executar comandos de um script.

Author: ```suggestion
O Console R, ou simplesmente Console, é por padrão o painel esquerdo ou inferior esquerdo do RStudio. Ele é a casa do "motor" de R. É aqui que os comandos são realmente executados e as saídas não gráficas e as mensagens de erro/aviso aparecem. Você pode digitar e executar comandos diretamente no Console, mas perceba que estes comandos não são salvos como são ao executar comandos de um script.
```

### Instalar e carregar{.unnumbered}

*As funções* estão contidas nos **pacotes** que podem ser baixados ("instalados") para seu computador a partir da Internet. Uma vez que um pacote é baixado, ele é armazenado em sua "biblioteca" (do inglês, *library*). Você pode então acessar as funções que ele contém durante sua sessão R atual "carregando" este pacote.

*Pense em R como sua biblioteca pessoal*: Quando você baixa um pacote, sua biblioteca ganha um novo livro de funções, mas cada vez que você quiser usar uma função naquele livro, você deve pegar emprestado ("carregar") aquele livro de sua biblioteca.

Em resumo: para usar as funções disponíveis em um pacote R, duas etapas devem ser implementadas:

1)  O pacote precisa ser **instalado** (uma vez), *e*  

2)  O pacote precisa ser **carregado** (a cada sessão de R que abrir)

> ####Sua biblioteca {.unnumbered}

Author: ```suggestion
#### Sua biblioteca {.unnumbered}
```

### R markdown {.unnumbered}

Um script R markdown é um tipo de script R no qual o script em si *resulta* um documento de saída (PDF, Word, HTML, Powerpoint, etc.). Estas são ferramentas incrivelmente úteis e versáteis, freqüentemente usadas para criar relatórios dinâmicos e automatizados. Mesmo este website e manual são produzidos com um script  R markdown!

Vale a pena notar que usuários iniciantes de R também podem usar R Markdown - não se intimidem! Para saber mais, consulte a página do manual nos documentos [Relatórios com R Markdown].

<!-- ======================================================= -->

### Notebooks R {.sem numeração}

Não há diferença entre escrever em um Rmarkdown versus um caderno R. Entretanto, a execução do documento difere ligeiramente. Consulte este [site](http://uc-r.github.io/r_notebook) para obter mais detalhes.

<!-- ======================================================= -->

### Shiny

Os aplicativos/websites shiny estão contidas em um script, que deve ser chamado de `app.R`. Este arquivo tem três componentes:

> 1) Uma interface de usuário (ui)  

Author: ```suggestion
1) Uma interface de usuário (ui)  \
```

> 2) Uma função de servidor  

Author: ```suggestion
2) Uma função de servidor  \
```

3) Uma chamada para a função `shinyApp'.  

Veja a página do manual em [Dashboards com Shiny], ou este tutorial online: [Tutorial com shiny](https://shiny.rstudio.com/tutorial/written-tutorial/lesson1/)

> *Nos tempos mais antigos, o arquivo acima era dividido em dois arquivos (`ui.R` e `server.R`)*

Author: ```suggestion
*Nos tempos mais antigos, o arquivo acima era dividido em dois arquivos (`ui.R`  e `server.R`)*
```

**Colunas**

As colunas em um conjunto de dados também são objetos e podem ser definidas, sobrescritas e criadas conforme descrito abaixo na seção Colunas.

Você pode usar o operador de atribuição do R **base** para criar uma nova coluna. Abaixo, a nova coluna `bmi` (IMC - Índice de Massa Corporal, do inglês *Body Mass Index*) é criada, e para cada linha o novo valor é resultado de uma operação matemática sobre o valor da linha nas colunas `wt_kg` e `ht_cm`.

```{r, eval=F}
# criar uma nova colna bmi (que é o valor de IMC) usando a sintaxe do R base
linelist$bmi <- linelist$wt_kg / (linelist$ht_cm/100)^2
```

No entanto, neste manual, enfatizamos uma abordagem diferente para definir colunas, que usa a função `mutate()` do pacote **dplyr** e *piping* com o operador pipe (`%>%`). A sintaxe é mais fácil de ler e há outras vantagens explicadas na página em [Limpeza de dados e principais funções ]. Você pode ler mais sobre *tubulação* na seção Tubulação abaixo.

```{r, eval=F}
# criar uma nova colna bmi (que é o valor de IMC) usando a sintaxe do dplyr
linelist <- linelist %>% 
  mutate(bmi = wt_kg / (ht_cm/100)^2)
```

<!-- ======================================================= -->

> ### Estruture de um objeto {.unnumbered}

Author: ```suggestion
### Estrutura de um objeto {.unnumbered}
```

| Numeric    | Números que **podem incluir decimais**. Se estiverem dentro de aspas, serão considerados como caractere                                                                      | 23.1 or 14                                                                                            |
+------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-------------------------------------------------------------------------------------------------------+
| Factor     | Fatores, são vatores que tem uma **ordem específica** ou hierarquia de valores                                                                                                                | Uma variável de status econômico com valores ordenados   |
+------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-------------------------------------------------------------------------------------------------------+

> | Date       | **Uma vez que o R é informado que um certo grupo de dados corresponde a Data**, esses valores só podem ser apresentados e manipulados de algumas maneiras especiais. Veja a página sobre [Trabalhando com datas] para mais informações                 | 2018-04-12 ou 15/3/1954 ou Wed 4 Jan 1980                                                             |

Author: ```suggestion
| Date       | **Uma vez que o R é informado que um certo grupo de dados corresponde a Data**, esses valores só podem ser apresentados e manipulados de algumas maneiras especiais. Veja a página sobre [Trabalhando com datas] para mais informações                 | 2018-04-12 ou 15/3/1954 ou Qua 4 Jan 1980                                                             |
```

+------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-------------------------------------------------------------------------------------------------------+
| Logical    | Valores lógicos que precisam um dos dois valores especiais TRUE (verdadeiro) ou FALSE (note que eles **não** não "TRUE" e "FALSE" entre aspas)                                                               | TRUE ou FALSE                                                                                         |
+------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-------------------------------------------------------------------------------------------------------+

> | data.frame | A data frame is how R stores a **typical dataset**. It consists of vectors (columns) of data bound together, that all have the same number of observations (rows).                      | The example AJS dataset named `linelist_raw` contains 68 variables with 300 observations (rows) each. |

Author: ```suggestion
| data.frame | Uma data frame é como o R armazena uma **base de dados típica**. Consiste de vetores (volunas) de dados colocados juntas, que tês o mesmo número de observações. | O exemplo da base de dados AJS nomeado  `linelist_raw` contém 68 variáveis com 300 observações (linhas) cada. |
```

+------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-------------------------------------------------------------------------------------------------------+

> | tibble     | tibble é uma variação do data frame, a principal diferença operacional é que eles imprimem mais bem no console (exibem as primeiras 10 linhas e apenas as colunas que cabem na tela)| Qualquer data frame, lista ou matriz pode ser convertido em um tibble com `as_tibble()`  |

Author: ```suggestion
| tibble     | tibble é uma variação do data frame, a principal diferença operacional é que eles printam melhor no console (exibem as primeiras 10 linhas e apenas as colunas que cabem na tela)| Qualquer data frame, lista ou matriz pode ser convertido em um tibble com `as_tibble()`  |
```

+------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-------------------------------------------------------------------------------------------------------+

> | list       | Uma lista é como um vetor, mas contém outros objetos que podem ser outras classes diferentes| Uma lista pode conter um único número, um dataframe, um vetor e até outra lista dentro dele!    |

Author: ```suggestion
| list       | Uma lista é como um vetor, mas contém outros objetos que podem ser de outras classes diferentes| Uma lista pode conter um único número, um dataframe, um vetor e até outra lista dentro dele!    |
```

class(linelist$age)     # classe deve ser numérica 

class(linelist$gender)  # classe deve ser caractere 
```

Às vezes, uma coluna será convertida automaticamente para uma classe diferente pelo R. Cuidado com isso! Por exemplo, se você tiver um vetor ou coluna de números, mas um valor de caractere for inserido... a coluna inteira mudará para caractere de classe.

```{r}
num_vector <- c(1,2,3,4,5) # define um vetor só de números
class(num_vector)          # este vetor é da classe numérico
num_vector[3] <- "three"   # converte o terceiro elemento para um caractere
class(num_vector)          # o vetor agora é classe caractere
```

Um exemplo comum disso é quando se manipula um data frame para exibir uma tabela - se você fizer uma linha total e tentar colar/colar porcentagens na mesma célula dos números (por exemplo, `23 (40%)`), toda a coluna numérica acima será convertida em caractere e não poderá mais ser utilizada para cálculos matemáticos.** Algumas vezes, será necessário converter objetos ou colunas em outra classe.**

+------------------+---------------------------------------------------------------------------------------+

> | Funço        | Ação                                                                                |

Author: ```suggestion
| Função        | Ação                                                                                |
```

| `as.integer()`   |Converte para a classe inteiro                                                             |
+------------------+---------------------------------------------------------------------------------------+
| `as.Date()`      | Converte para a classe Data - Nota: Veja a seção sobre [datas](#dates) para detalhes      |
+------------------+---------------------------------------------------------------------------------------+
| `factor()`       | Converte para a classe fator - Nota: Redefinir a ordem dos níveis requer argumentos extras |
+------------------+---------------------------------------------------------------------------------------+

Da mesma forma, existem as funções do R **base** para verificar se um objeto é de uma classe específica, como `is.numeric()`, `is.character()`, `is.double()`, `is.factor()`, `is.integer()`

Aqui está [mais material on-line sobre classes e estruturas de dados em R](https://swcarpentry.github.io/r-novice-inflammation/13-supp-data-structures/).

<!-- ======================================================= -->

### Colunas/Variáveis (`$`) {.unnumbered}

> **Uma coluna em um quadro de dados é tecnicamente um "vetor" (ver tabela acima)** - uma série de valores que devem ser todos da mesma classe (tanto caracter, numérico, lógico, etc.).

Author: ```suggestion
**Uma coluna em um dataframe é tecnicamente um "vetor" (ver tabela acima)** - uma série de valores que devem ser todos da mesma classe (tanto caracter, numérico, lógico, etc.).
```

```

+------------------------------------------------------------------------------------------------+--------------------------------------------+
| Critérios no exemplo acima                                                                      | Valor resultando na nova variável  "case_def" |
+================================================================================================+============================================+
| Se o valor das variáveis `rdt_result` e `other_cases_in_home` estiverem faltando                  | `NA` (faltante)                             |
+------------------------------------------------------------------------------------------------+--------------------------------------------+
| Se o valor em `rdt_result` for "Positive"                                                     | "Confirmed"                                |
+------------------------------------------------------------------------------------------------+--------------------------------------------+

> | If the value in `rdt_result` is NOT "Positive" AND the value in `other_cases_in_home` is "Yes" | "Probable"                                 |

Author: ```suggestion
| Se o valor em  `rdt_result` NÃO for "Positivo" E o valor em `other_cases_in_home` for "Yes" | "Probable"                                 |
```

<!-- ======================================================= -->

> ###Valores faltantes {.unnumbered}

Author: ```suggestion
### Valores faltantes {.unnumbered}
```

Em R, os valores ausentes são representados pelo valor especial `NA` (um valor "reservado") (letras maiúsculas N e A - não entre aspas). Se você importar dados que registram dados ausentes de outra maneira (por exemplo, 99, "Missing" ou .), convém recodificar esses valores para `NA`. Como fazer isso é abordado na página [Importação e exportação].

**Para testar se um valor é `NA`, use a função especial `is.na()`**, que retorna `TRUE` ou `FALSE`.

```{r basics_operators_missing}
rdt_result <- c("Positive", "Suspected", "Positive", NA)   # dois casos positivos, um suspeito e um desconhecido 
is.na(rdt_result)  # Testa se o valor de  rdt_result é NA
```

> Leia mais sobre valores ausentes, infinitos, `NULL` e impossíveis na página em [Valores faltantes]. Saiba como converter valores ausentes ao importar dados na página em [Importação e exportação].

Author: ```suggestion
Leia mais sobre valores ausentes, infinitos, `NULL` e impossíveis na página em [Campos em branco/faltantes]. Saiba como converter valores ausentes ao importar dados na página em [Importação e exportação].
```

<!-- ======================================================= -->

> ### Matemática and estatística {.unnumbered}

Author: ```suggestion
### Matemática e estatística {.unnumbered}
```

Todos os operadores e funções nesta página estão disponíveis automaticamente usando o pacote R **base**.

#### Operadores mmatemáticos {.unnumbered}

Estes são frequentemente usados para realizar adição, divisão, para criar novas colunas, etc. Abaixo estão os operadores matemáticos comuns em R. Colocar ou não os espaços ao redor dos operadores não é importante.

> | Propósit            | Exemplo no R |

Author: ```suggestion
| Propósito            | Exemplo no R |
```

|---------------------|--------------|
| adição              | 2 + 3        |
| subtração           | 2 - 3        |
| multiplicação       | 2 \* 3       |

> | divisão  b          | 30 / 5       |

Author: ```suggestion
| divisão             | 30 / 5       |
```

| exponencial                 | exponent(x)                           |
| logaritmo natural/neperiano | log(x)                                |
| logaritmo na base 10        | log10(x)                              |
| logaritmo na base 2         | log2(x)                               |

Nota: para `round()` o `digits =` especifica o número de casas decimais colocadas. Use `signif()` para arredondar para um número de algarismos significativos.

#### Notação científica {.unnumbered}

A probabilidade de notação científica ser usada depende do valor da opção `scipen`.

Da documentação de `?options`: scipen é uma penalidade a ser aplicada ao decidir imprimir valores numéricos em notação fixa ou exponencial. Os valores positivos tendem para a notação fixa e os negativos para a notação científica: a notação fixa será preferida, a menos que seja mais do que os dígitos 'scipen' mais largos.

Se estiver definido para um número baixo (por exemplo, 0), ele estará sempre "ligado". Para "desativar" a notação científica em sua sessão R, defina-a como um número muito alto, por exemplo:

```{r, eval=F}
# desligar a notação científica
options(scipen=999)
```

#### Arredondando {.unnumbered}

> [***PERIGO:*** `round()` usa "arredondamento do banqueiro" que arredonda de .5 somente se o número superior for par. Use `round_half_up()` do **zelador** para arredondar consistentemente metades para o número inteiro mais próximo. Veja [esta explicação](https://cran.r-project.org/web/packages/janitor/vignettes/janitor.html#explore-records-with-duplicated-values-for-specific-combinations-of-variables-with-get_dupes)]{style="color: red;"}

Author: ```suggestion
[***PERIGO:*** `round()` usa "arredondamento do banqueiro" que arredonda de .5 somente se o número superior for par. Use `round_half_up()` do **janitor** para arredondar consistentemente metades para o número inteiro mais próximo. Veja [esta explicação](https://cran.r-project.org/web/packages/janitor/vignettes/janitor.html#explore-records-with-duplicated-values-for-specific-combinations-of-variables-with-get_dupes)]{style="color: red;"}
```

```{r}
# use a função de arredondamento adequanda para o seu trabalho 
round(c(2.5, 3.5))

janitor::round_half_up(c(2.5, 3.5))
```

#### Funções estatísticas {.unnumbered}

[***CUIDADO:*** As funções abaixo incluirão, por padrão, valores ausentes nos cálculos. Valores ausentes resultarão em uma saída de `NA`, a menos que o argumento `na.rm = TRUE` seja especificado. Isso pode ser escrito de forma abreviada como `na.rm = T`.]{style="color: orange;"}

> | ObjetiVO                   | Função             |

Author: ```suggestion
| Objetivo                  | Função             |
```
